### PR TITLE
GOVUKAPP-685: Analytics Consent

### PR DESCRIPTION
**Pull request by TimGJD**
[Figma Link](https://www.figma.com/design/N4v2GPQhwZ7DwhqucyJqlu/GOV.UK-Apps-and-notifications?node-id=12983-28164&t=9Zq4aR8McUWqPgwh-4)

[Jira Link](https://govukverify.atlassian.net/browse/GOVUKAPP-685)

Light Mode
image

Dark Mode
image

### CODE REVIEW

File: Production/govuk_ios/SupportingFiles/AppDelegate.swift [hunk 1]
```diff
@@ -8,7 +8,6 @@ class AppDelegate: UIResponder,
                      didFinishLaunchingWithOptions
                      launchOptions: [UIApplication.LaunchOptionsKey: Any]? = nil) -> Bool {
         analyticsService.configure()
-        analyticsService.setAcceptedAnalytics(accepted: true)
```
Comment (TimGJD) [inline Production/govuk_ios/SupportingFiles/AppDelegate.swift:11]: Highlighting this removal as seems this function call was initially added to always enable analytics for dev/test purposes as there was previously no other way of enabling? 
Removing this now that we can enable and disable analytics on app startup using Analytics Consent screen.

File: Production/govuk_ios/Extensions/UIKit/URLOpening.swift [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import UIKit
+
+protocol URLOpening {
+    func canOpenURL(_ url: URL) -> Bool
+    func open(_ url: URL,
+              options: [UIApplication.OpenExternalURLOptionsKey: Any],
+              completionHandler completion: ((Bool) -> Void)?
+    )
```
Comment (batesyuk) [inline Production/govuk_ios/Extensions/UIKit/URLOpening.swift:8]: Can you knock this up to line 7, please? I know this seems a little off from other areas, I can talk you though the madness though

File: Tests/govuk_ios/govuk_ios_unit_tests/Mocks/Extensions/MockURLOpening.swift [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+@testable import govuk_ios
+
+class MockURLOpening: URLOpening {
+    var _stubbedCanOpenUrl: Bool = false
+    func canOpenURL(_ url: URL) -> Bool {
+        return _stubbedCanOpenUrl
+    }
+
+    var _receivedOpenCompletion: Bool = false
+    func open(_ url: URL, options: [UIApplication.OpenExternalURLOptionsKey : Any], completionHandler completion: ((Bool) -> Void)?) {
```
Comment (batesyuk) [inline Tests/govuk_ios/govuk_ios_unit_tests/Mocks/Extensions/MockURLOpening.swift:12]: Can you please knock these down to new lines?

File: Tests/govuk_ios/govuk_ios_unit_tests/Specs/Coordinators/AnalyticsConsentCoordinatorTests.swift [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            analyticsService: MockAnalyticsService(),
+            dismissAction: {
+                expectation.fulfill()
+            }
+        )
+        sut.start()
+        wait(for: [expectation], timeout: 1)
+    }
+
+    func test_start_hasNotSeenAnalyticsConsent_doesNotCallDismiss() throws {
```
Comment (batesyuk) [inline Tests/govuk_ios/govuk_ios_unit_tests/Specs/Coordinators/AnalyticsConsentCoordinatorTests.swift:23]: You can do an inverse expectation here to make sure that the XCTFail isn't being called after the test is considered complete.

File: Production/govuk_ios/Extensions/UIKit/URLOpening.swift (hunk 2 above)
Comment (batesyuk) [inline Production/govuk_ios/Extensions/UIKit/URLOpening.swift:8]: Can you move this `)` to line 7, please?

File: Tests/govuk_ios/govuk_ios_unit_tests/Mocks/Extensions/MockURLOpening.swift (hunk 3 above)
Comment (batesyuk) [inline Tests/govuk_ios/govuk_ios_unit_tests/Mocks/Extensions/MockURLOpening.swift:12]: new lines for params here, please.

File: Tests/govuk_ios/govuk_ios_unit_tests/Specs/Coordinators/AnalyticsConsentCoordinatorTests.swift [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        let analyticsConsentService = MockAnalyticsConsentService()
+        analyticsConsentService._stubbedHasSeenAnalyticsConsent = false
+        let mockNavigationController = UINavigationController()
+        let sut = AnalyticsConsentCoordinator(
+            navigationController: mockNavigationController,
+            analyticsConsentService: analyticsConsentService,
+            analyticsService: MockAnalyticsService(),
+            dismissAction: {
+                XCTFail()
```
Comment (batesyuk) [inline Tests/govuk_ios/govuk_ios_unit_tests/Specs/Coordinators/AnalyticsConsentCoordinatorTests.swift:32]: You could do an inverse expectation here to make sure that the `XCTFail` isn't being called after the test has completed

File: Tests/govuk_ios/govuk_ios_unit_tests/Mocks/Extensions/MockURLOpening.swift (hunk 3 above)
Comment (TimGJD) [reply Tests/govuk_ios/govuk_ios_unit_tests/Mocks/Extensions/MockURLOpening.swift:14]: Done

Comment (TimGJD) [reply Tests/govuk_ios/govuk_ios_unit_tests/Mocks/Extensions/MockURLOpening.swift:14]: Done

File: Production/govuk_ios/Extensions/UIKit/URLOpening.swift (hunk 2 above)
Comment (TimGJD) [reply Production/govuk_ios/Extensions/UIKit/URLOpening.swift]: Done

File: Tests/govuk_ios/govuk_ios_unit_tests/Specs/Coordinators/AnalyticsConsentCoordinatorTests.swift (hunk 4 above)
Comment (TimGJD) [reply Tests/govuk_ios/govuk_ios_unit_tests/Specs/Coordinators/AnalyticsConsentCoordinatorTests.swift]: Done

File: Tests/govuk_ios/govuk_ios_unit_tests/Specs/Coordinators/AnalyticsConsentCoordinatorTests.swift (hunk 5 above)
Comment (TimGJD) [reply Tests/govuk_ios/govuk_ios_unit_tests/Specs/Coordinators/AnalyticsConsentCoordinatorTests.swift]: Done

File: Production/govuk_ios/Extensions/UIKit/URLOpening.swift (hunk 2 above)
Comment (TimGJD) [reply Production/govuk_ios/Extensions/UIKit/URLOpening.swift]: Done

**Review by batesyuk — APPROVED** [review]

**Review by batesyuk — APPROVED** [review]

**Pull request merged**